packages/langium/src/index/index-manager.ts

```diff
@@ -0,0 +1,125 @@
+import { WorkspaceFolder } from 'vscode-languageserver';
+import { LangiumDocument, LangiumDocuments } from '../documents/document';
+import { LanguageMetaData } from '../grammar/language-meta-data';
+import { AstNodeDescription } from '../references/scope';
+import { LangiumServices } from '../services';
+import { AstNodeDescriptionProvider, ReferenceDescription, ReferenceDescriptionProvider } from './ast-descriptions';
+import { AstNode } from '../syntax-tree';
+import { stream, Stream } from '../utils/stream';
+import { getDocument } from '../utils/ast-util';
+import { AstReflection } from '..';
+
+export interface IndexManager {
+    initializeRoot(rootUri: string): void;
+    initializeWorkspace(folders: WorkspaceFolder[] | null): void;
+    update(document: LangiumDocument): void;
+    /* Use streams? */
+    allElements(referenceType?: string): Stream<AstNodeDescription>;
+    findAllReferences(targetNode: AstNode, astNodePath: string): Stream<ReferenceDescription>;
+}
+
+export class DefaultIndexManager implements IndexManager {
+
+    protected readonly langiumDocuments: () => LangiumDocuments;
+    protected readonly astNodeDescriptionProvider: () => AstNodeDescriptionProvider;
+    protected readonly referenceDescriptionProvider: () => ReferenceDescriptionProvider;
+    protected readonly languageMetaData: LanguageMetaData;
+    protected readonly astReflection: AstReflection;
+
+    simpleIndex: Map<string, AstNodeDescription[]> = new Map<string, AstNodeDescription[]>();
+    referenceIndex: Map<string, ReferenceDescription[]> = new Map<string, ReferenceDescription[]>();
+
+    constructor(services: LangiumServices) {
+        this.astReflection = services.AstReflection;
+        this.languageMetaData = services.LanguageMetaData;
+        this.langiumDocuments = () => services.documents.LangiumDocuments;
+        this.astNodeDescriptionProvider = () => services.index.AstNodeDescriptionProvider;
+        this.referenceDescriptionProvider = () => services.index.ReferenceDescriptionProvider;
+    }
+
+    documentDescriptions(): ReadonlyMap<string, ReferenceDescription[]> {
+        return this.referenceIndex;
+    }
+
+    findAllReferences(targetNode: AstNode, astNodePath: string): Stream<ReferenceDescription> {
+        const targetDocUri = getDocument(targetNode).textDocument.uri;
+        const result: ReferenceDescription[] = [];
+        this.referenceIndex.forEach((docRefs: ReferenceDescription[]) => {
+            docRefs.forEach((refDescr)=> {
+                if(refDescr.targetUri === targetDocUri && refDescr.targetPath === astNodePath) {
+                    result.push(refDescr);
+                }
+            });
+        });
+        return stream(result);
+    }
+
+    allElements(referenceType?: string): Stream<AstNodeDescription> {
+        return stream(Array.from(this.simpleIndex.values()).flat().filter(e => referenceType ? this.astReflection.isSubtype(e.type, referenceType) : true));
```

Comment:
More efficient:
```typescript
return stream(this.simpleIndex.values()).flat().filter(...);
```

Reply:
Is it really more efficient though? If we iterate more than once over the returned stream, we need to flatten and filter everything on every call, don't we? As everything is lazily evaluated.

Reply:
That's responsibility of the caller. If the caller knows they will iterate multiple times, they can convert it to an array first. If they iterate only once, it's more efficient to operate on the stream.

Reply:
>That's responsibility of the caller

> they can convert it to an array first

My impression is that we should probably keep the `allElements()` as it was by returning really all references as an array and add an additional API `elementsByType(referenceType?: string): Stream<>` or maybe even `elementsBy(filter: (node: AstNodeDescription) => boolean): Stream<AstNodeDescription>;`

Anyway, please do such kind of optimization in a separate Issue/PR.

Reply:
Ok, please create an issue then.